Mega menu fixes

wjhsf opened the pull request:
<!--- Provide a short summary of your changes in the Title field above -->
1. Category stale time was not being properly checked.
2. Category fetch time was being set on page load, rather than when it was fetched. (If on page load is the desired behavior, then `fetchTime` is a misnomer...)
3. Loaded category data was `Category | Error`, but we just want `Category`.
4. Accidental error when `root[itemsKey]` is undefined.

# Description

<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relevant context. -->

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

- (step1)

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

wjhsf committed:
Remove unnecessary catch clause.

wjhsf committed:
Update copyright.

wjhsf committed:
Don't return error when category is expected.

wjhsf committed:
Fix bug preventing cache invalidation.

wjhsf committed:
Set fetchTime when data is fetched, rather than every page load.

wjhsf committed:
Update comment.

wjhsf committed:
Convert promise chain to async/await for readability.

wjhsf commented on packages/template-retail-react-app/app/contexts/index.js:
```diff
@@ -40,65 +40,51 @@ export const CategoriesProvider = ({treeRoot = {}, children, locale}) => {
     const api = useCommerceAPI()
     const [root, setRoot] = useState({
-        // map over the server-provided cat
+        // Clone the server-provided category
```
I don't remember what `useState`'s internals are like. Is cloning `treeRoot` necessary or is it a vestige of when we mutated the data?

bfeister replied:
We still need to copy tree root just in terms of React internals and object observability, you want to mutate the whole object

wjhsf commented on packages/template-retail-react-app/app/contexts/index.js:
```diff
@@ -40,65 +40,51 @@ export const CategoriesProvider = ({treeRoot = {}, children, locale}) => {
-            window.localStorage.getItem(`${LOCAL_STORAGE_PREFIX}${id}-${locale}`)
-        )
-        if (storageItem || Date.now() < storageItem?.fetchTime + CAT_MENU_STALE_TIME) {
+        if (storageItem && Date.now() < storageItem.fetchTime + CAT_MENU_STALE_TIME) {
```
`||` meant that we would _always_ return the stored data, regardless of staleness.

bfeister replied:
Gotcha, my fault on missing this, I missed the `||` which was the relevant bit

wjhsf commented on packages/template-retail-react-app/app/contexts/index.js:
```diff
@@ -40,65 +40,51 @@ export const CategoriesProvider = ({treeRoot = {}, children, locale}) => {
                 levels
             }
         })
+        res.loaded = true
```
There used to be a lot of `loaded` props floating around. Is this one still necessary?

bfeister replied:
Yes, it's used in the `list-menu/index.jsx` component

wjhsf committed:
Guard against missing category items.

wjhsf commented on packages/template-retail-react-app/app/contexts/index.js:
```diff
@@ -40,65 +40,51 @@ export const CategoriesProvider = ({treeRoot = {}, children, locale}) => {
     }
 
-    useEffect(() => {
+    useEffect(async () => {
```
Does `useEffect` behave differently when the callback returns `Promise<void>` instead of `void`? I don't think so, so I switched from promise chains to async/await. This is purely for readability, though, so can be reverted if there are unintended consequences.

bfeister replied:
Did you test your change locally to check manually for expected behavior?

wjhsf replied:
Yeah, but this feels like the kind of thing that would have sneaky, hidden side effects, so I wanted to make sure.

bfeister replied:
Let's revert this.

> the first argument of useEffect is supposed to be a function that returns either nothing (undefined) or a function (to clean up side effects). But an async function returns a Promise, which can't be called as a function! It's simply not what the useEffect hook expects for its first argument.

yunakim714 replied:
+1, we do not want to have `useEffect(async () => {`

wjhsf committed:
Move comment inside code block.

wjhsf commented on packages/template-retail-react-app/app/contexts/index.js:
```diff
@@ -40,65 +40,53 @@ export const CategoriesProvider = ({treeRoot = {}, children, locale}) => {
-            })
-            .catch((err) => {
-                throw err
+        const promises = root?.[itemsKey]?.map(async (cat) => await fetchCategoryNode(cat?.id, 2))
+        if (promises) {
```
Does the data from the API potentially lack `[itemsKey]`? What would be the consequence of having it default to an empty array if it is missing (way up in `useState`, not here)? We'd have to change relevant checks from `root?.[itemsKey]` to `root[itemsKey].length`, but how many of those are there? Are there other consequences that I'm not thinking of?

bfeister replied:
`itemsKey` is a default constant set in the app, and the API will always respond with `categories` (which is the hardcoded `itemsKey`

bfeister replied:
My preference is to keep optional chaining bits simply because they introduce fault tolerance that reduces the chance of customers seeing thrown errors. Perhaps the API later changes their response schema, it does not cause any harm to be "lucky" and prepared for such a thing even if it never happens

wjhsf commented on packages/template-retail-react-app/app/contexts/index.js:
```diff
@@ -40,65 +40,53 @@ export const CategoriesProvider = ({treeRoot = {}, children, locale}) => {
             }
         })
+        res.loaded = true
+        window.localStorage.setItem(
```
Moved this to inside the fetch function, so that fetch time is set on fetch, rather than on every page load.

bfeister reviewed (changes requested):
This looks good aside from the `useEffect(async () =>` bit which needs to be reverted. Thanks for the catch on STALE_TIME

wjhsf committed:
Change useEffect back to promise chain, not async/await.

useEffect expects the callback to return nothing or a function;
returning a promise could break things.

bfeister commented on packages/template-retail-react-app/app/contexts/index.js:
```diff
@@ -40,65 +40,54 @@ export const CategoriesProvider = ({treeRoot = {}, children, locale}) => {
+        const promises = root?.[itemsKey]?.map(async (cat) => await fetchCategoryNode(cat?.id, 2))
+        if (promises) {
+            // TODO: Error handling when fetching data fails.
+            // Possibly switch to .allSettled to show partial data?
```
`.allSettled` could be good here. Good suggestion
